cpp/src_prims/distance/pairwise_distance_base.cuh

```diff
@@ -0,0 +1,182 @@
+          typename EpilogueLambda, typename FinalLambda,
+          typename BaseClass =
+            raft::linalg::Contractions_NT<DataT, IdxT, Policy>>
+struct PairwiseDistances : public BaseClass {
+ private:
+  typedef Policy P;
+  const DataT* xn;
+  const DataT* yn;
+  DataT* sxNorm;
+  DataT* syNorm;
+  OutT* dOutput;
+  char* smem;
+  CoreLambda core_op;
+  EpilogueLambda epilog_op;
+  FinalLambda fin_op;
+
+  AccT acc[P::AccRowsPerTh][P::AccColsPerTh];
+
+ public:
+  // Constructor for euclidean expanded
+  DI PairwiseDistances(const DataT* _x, const DataT* _y, IdxT _m, IdxT _n,
+                       IdxT _k, const DataT* _xn, const DataT* _yn,
+                       OutT* _dOutput, char* _smem, CoreLambda _core_op,
+                       EpilogueLambda _epilog_op, FinalLambda _fin_op)
+    : BaseClass(_x, _y, _m, _n, _k, _smem),
+      sxNorm((DataT*)_smem),
+      syNorm(&(sxNorm[P::Mblk])),
+      xn(_xn),
+      yn(_yn),
+      dOutput(_dOutput),
+      smem(_smem),
+      core_op(_core_op),
+      epilog_op(_epilog_op),
+      fin_op(_fin_op) {}
+
+  // Constructor for euclidean unexpanded
+  DI PairwiseDistances(const DataT* _x, const DataT* _y, IdxT _m, IdxT _n,
+                       IdxT _k, OutT* _dOutput, char* _smem,
+                       CoreLambda _core_op, EpilogueLambda _epilog_op,
+                       FinalLambda _fin_op)
+    : BaseClass(_x, _y, _m, _n, _k, _smem),
+      dOutput(_dOutput),
+      smem(_smem),
+      core_op(_core_op),
+      epilog_op(_epilog_op),
+      fin_op(_fin_op) {}
+
+  DI void run() {
+    prolog();
+    loop();
+    epilog();
+  }
+
+ private:
+  DI void prolog() {
+    this->ldgXY(0);
+#pragma unroll
+    for (int i = 0; i < P::AccRowsPerTh; ++i) {
+#pragma unroll
+      for (int j = 0; j < P::AccColsPerTh; ++j) {
+        acc[i][j] = BaseClass::Zero;
+      }
+    }
+    this->stsXY();
+    __syncthreads();
+    this->pageWr ^= 1;
+  }
+
+  DI void loop() {
+    for (int kidx = P::Kblk; kidx < this->k; kidx += P::Kblk) {
+      this->ldgXY(kidx);
+      accumulate();  // on the previous k-block
+      this->stsXY();
+      __syncthreads();
+      this->pageWr ^= 1;
+      this->pageRd ^= 1;
+    }
+    accumulate();  // last iteration
+  }
+
+  DI void accumulate() {
+#pragma unroll
+    for (int ki = 0; ki < P::Kblk; ki += P::Veclen) {
+      this->ldsXY(ki);
+#pragma unroll
+      for (int i = 0; i < P::AccRowsPerTh; ++i) {
+#pragma unroll
+        for (int j = 0; j < P::AccColsPerTh; ++j) {
+#pragma unroll
+          for (int v = 0; v < P::Veclen; ++v) {
+            core_op(acc[i][j], this->regx[i][v], this->regy[j][v]);
+          }
+        }
+      }
+    }
+  }
+
+  DI void epilog() {
+    epilog_op(acc, sxNorm, syNorm);
+
+    if (Sqrt) {
+#pragma unroll
+      for (int i = 0; i < P::AccRowsPerTh; ++i) {
+#pragma unroll
+        for (int j = 0; j < P::AccColsPerTh; ++j) {
+          acc[i][j] = raft::mySqrt(acc[i][j]);
+        }
+      }
+    }
```

Comment:
we could potentially have `epilog_op` do this, right? IOW, if sqrt is needed, let the final distance metrics pass an `epilog_op` which does sqrt inside. What say?

Reply:
Yes I think that makes sense as only L2 exp/unexp need sqrt and no other distance metric. I will move this to epilog_op of L2 only.